packages/unified-selection/src/unified-selection/HiliteSetProvider.ts

```diff
@@ -0,0 +1,338 @@
+    const elementKeys = map.get("element") ?? [];
+
+    if (this._hasFunctionalSchema) {
+      const functionalGeometricElements = new Set(await this.getHilitedFunctionalElements(map));
+      geometricElementKeySet = new Set([...geometricElementKeySet, ...functionalGeometricElements]);
+    }
+
+    const geometricElementKeys = Array.from(geometricElementKeySet);
+
+    if (groupInformationElementKeys.length === 0 && geometricElementKeys.length === 0 && elementKeys.length === 0) {
+      return [];
+    }
+
+    const bindings: ECSqlBinding[] = [
+      { type: "idset", value: groupInformationElementKeys },
+      { type: "idset", value: elementKeys },
+      { type: "idset", value: geometricElementKeys },
+    ];
+    const query = `WITH RECURSIVE
+                    GroupMembers (ECInstanceId, ECClassId) AS (
+                      SELECT r.TargetECInstanceId, r.TargetECClassId FROM BisCore.GroupInformationElement s
+                        JOIN BisCore.ElementGroupsMembers r ON r.SourceECInstanceId = s.ECInstanceId
+                      WHERE InVirtualSet(?, s.ECInstanceId)
+                    ),
+                    GroupGeometricElements (ECInstanceId, ECClassId) AS (
+                      SELECT ECInstanceId, ECClassId FROM GroupMembers
+                      UNION ALL
+                      SELECT r.TargetECInstanceId, r.TargetECClassId FROM GroupGeometricElements s
+                        JOIN BisCore.ElementOwnsChildElements r ON r.SourceECInstanceId = s.ECInstanceId
+                    ),
+                    ElementGeometricElements (ECInstanceId, ECClassId) AS (
+                      SELECT ECInstanceId, ECClassId FROM BisCore.Element WHERE InVirtualSet(?, ECInstanceId)
+                      UNION ALL
+                      SELECT r.TargetECInstanceId, r.TargetECClassId FROM ElementGeometricElements s
+                        JOIN BisCore.ElementOwnsChildElements r ON r.SourceECInstanceId = s.ECInstanceId
+                    ),
+                    MergedElements (ECInstanceId, ECClassId) AS (
+                      SELECT ECInstanceId, ECClassId FROM BisCore.GeometricElement WHERE InVirtualSet(?, ECInstanceId)
+                      UNION
+                      SELECT ECInstanceId, ECClassId FROM GroupGeometricElements WHERE ECClassId IS (BisCore.GeometricElement)
+                      UNION
+                      SELECT ECInstanceId, ECClassId FROM ElementGeometricElements WHERE ECClassId IS (BisCore.GeometricElement)
+                    ),
+                    GeometricElementGeometricElements (ECInstanceId, ECClassId) AS (
+                      SELECT ECInstanceId, ECClassId FROM MergedElements
+                      UNION ALL
+                      SELECT r.TargetECInstanceId, r.TargetECClassId FROM GeometricElementGeometricElements s
+                        JOIN BisCore.ElementOwnsChildElements r ON r.SourceECInstanceId = s.ECInstanceId
+                    )
```

Comment:
It was needed to recurse through passed geometric element children, however I split that into a separate table to avoid traversing the same elements again.